orttraining/orttraining/core/framework/pipeline.cc

```diff
@@ -347,6 +351,31 @@ void PipelineScheduler::CreateFullSchedule() {
   }
 }
 
+void PipelineScheduler::MapStageIdToMpiRank() {
+  for (int t = 0; static_cast<size_t>(t) < compute_commute_table_.size(); ++t) {
+    for (int s = 0; s < num_stages_; ++s) {
+      // Slot at time t and pipeline stage s. It's a collection of tasks.
+      auto& slot = compute_commute_table_[t][s];
+      for (int k = 0; k < slot.Size(); ++k) {
+        auto& task = slot[k];
+        if (!task.IsCommute()) {
+          continue;
+        }
+
+        // task.this_rank is actually the stage ID.
+        // task.this_rank=0 means this stage is the first pipeline stage.
+        // If data parallel is enabled, we may have multiple pipeline parallel groups.
+        // The code below maps stage ID to MPI's world rank.
+        task.this_rank = DistributedRunContext::GetRanks(WorkerGroupType::ModelParallel)[task.this_rank];
```

Comment:
Yes. If we have two DP groups, `this_rank` should be different for the same pipeline stage in different DP group. Otherwise, the Send/Recv may communicate data with wrong destination.

---
In reply to: [532604453](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r532604453) [](ancestors = 532604453)